New collation format

jannikluhn opened the pull request:
This implements the new collation format described in #420.

I deviated from the description slightly by removing the `period` field for now (see https://github.com/ethereum/py-evm/issues/420#issuecomment-372554339) and by replacing `blobs` with `body`. I think this makes sense at least for phase 1 because we don't care yet how the blobs will be encoded.

As expected this breaks a lot of tests (basically all that have anything to do with sharding). I'm not sure what the best way to deal with this is. My idea would simply be to remove everything that deals with transaction execution (this should include all failing tests), which isn't included in phase 1 anyway. Basically everything from `ShardingVM` and below. When we progress to stage 2 we can recover everything from the git history.

#### Cute Animal Picture

![2](https://user-images.githubusercontent.com/29854669/37348964-a0ff056a-26d5-11e8-9b04-83e272b7261e.jpg)

pipermerriam commented on evm/rlp/headers.py:
```diff
@@ -179,83 +179,135 @@ def create_execution_context(self, prev_hashes):
+            int_to_bytes32(self.proposer_bid),
+            self.proposer_signature,
+        ])
+        assert len(encoded) == self.smc_encoded_size
```
We have tended to avoid assertions in runtime code as they are [easy to disable](https://stackoverflow.com/questions/1273211/disable-assertions-in-python).  Preference is to raise explicit exceptions like `ValueError` in cases like this.

pipermerriam commented on evm/rlp/headers.py:
```diff
@@ -179,83 +179,135 @@ def create_execution_context(self, prev_hashes):
+            self.shard_id,
+        )
+
+    def to_bytes(self):
```
I'm confused how `to_bytes` and `rlp.encode(obj)` are related? or how `to_bytes` is used. Looking at the implementation of the `hash` property I would have expected it to be `keccak(rlp.encode(self))`.

Can we get a comment explaining the intent behind `to_bytes` as well as maybe another one in the `hash` implementation which explains why it uses `to_bytes` over `rlp.encode`.

jannikluhn replied:
`to_bytes` encode by padding each field to a fixed length and concatenating them. So it's less space efficient than RLP encoding, but much easier to decode for the SMC (which also has to calculate the hash).

I'm not very happy with the names `to_bytes` and `from_bytes` as it doesn't distinguish from RLP encoding, but I wanted to stay close to the existing implementation (there used to be `from_bytes`, I've added `to_bytes`). Maybe a name such as `encode_for_SMC` would clear things up already (but `decode_from_SMC` sounds weird).

pipermerriam replied:
👍 

I like `encode/decode_for/from_SMC` more than `to/from_bytes`.  Alternate names that come to mind.

- `encode_with_padding`
- `decode_padded`

jannikluhn replied:
@mhchia Which names would you prefer?

mhchia replied:
My apology that I just noticed this comment. 
`encode_with_padding`, `decode_padded` sounds good to me. `encode/decode_for/from_SMC` is easier for people to realize that this is a SMC specific encoding. I'm fine with both:)

pipermerriam commented on evm/rlp/headers.py:
```diff
@@ -179,83 +179,135 @@ def create_execution_context(self, prev_hashes):
-            start_index = idx * obj_size
-            field_bytes = header_bytes[start_index:(start_index + obj_size)]
+
+        start_indices = accumulate(lambda i, field: i + field[1], cls.smc_field_sizes, 0)
```
I've never used `accumulate` and now I ❤️ it!

pipermerriam commented on evm/rlp/headers.py:
```diff
@@ -179,83 +179,135 @@ def create_execution_context(self, prev_hashes):
+        start_indices = accumulate(lambda i, field: i + field[1], cls.smc_field_sizes, 0)
+        start_index = next(start_indices)
+        for field_name, field_type in cls.fields:
+            end_index = next(start_indices)
```
I think this loop can be changed as follows.

```
from cytoolz import sliding_window
field_bounds = sliding_window(2, start_indices)
for ((start_idx, end_idx), (field_name, field_type)) in zip(cls.fields, field_bounts):
    ...
```

Removes mutability as well as extracting the start/end index logic out of the loop body which I think improves readability some.

holgerd77 commented:
Hi, I did some stub implementation of collations for ethereumjs today heavily inspired by what you guys are doing here: https://github.com/ethereumjs/ethereumjs-collation

I was especially unsure about the field lengths for the different collation headers, this is what a default header initialization now look like:

```javascript
{ raw:
   [ <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00>,
     <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00>,
     <Buffer 56 e8 1f 17 1b cc 55 a6 ff 83 45 e6 92 c0 f8 6e 5b 48 e0 1b 99 6c ad c0 01 62 2f b5 e3 63 b4 21>,
     <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00>,
     <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00>,
     <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00>,
     <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00>,
     <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... > ],
  _fields:
   [ 'shardId',
     'parentHash',
     'chunkRoot',
     'period',
     'height',
     'proposerAddress',
     'proposerBid',
     'proposerSignature' ]
```

Would be great if you could have a short look if you immediately spot something wrong.

Another question: do you already have some RLP-serialized testing data for headers and/or bodies?

jannikluhn commented:
@holgerd77 Opened an issue [here](https://github.com/ethereumjs/ethereumjs-collation/issues/2).

We don't have any "official" test data yet unfortunately, but it's definitely on the roadmap.

jannikluhn commented:
@cburgdorf mypy on Travis is failing for this PR while it's passing on my machine:

```
evm/rlp/headers.py:254: error: Invalid type "eth_keys.keys"
evm/rlp/headers.py:257: error: keys.PrivateKey? has no attribute "public_key"
evm/rlp/headers.py:260: error: keys.PrivateKey? has no attribute "sign_msg_hash"
```

I think the reason for this is that the `eth_keys` package is not installed for the mypy run. Is that correct and if so what's the right place to add it? `dev_requirements.txt`?

cburgdorf commented:
@jannikluhn 

>I think the reason for this is that the eth_keys package is not installed for the mypy run

Nope, the dependency is installed (see the selected text on this screenshot from your travis build)

![image](https://user-images.githubusercontent.com/521109/38458161-38f9820e-3a9b-11e8-89d2-18a769538f34.png)

I haven't really looked into `eth_keys` but it appears that `PrivateKey` isn't exported on that level. If you change the import to this

```
from eth_keys.datatypes import (
    PrivateKey
)
```

And then on line `254` you replace `keys.PrivateKey` with `PrivateKey` the error should disappear.

jannikluhn commented:
Thanks a lot, @cburgdorf, importing `PrivateKey` directly fixed it.

Should be ready for review now.

jannikluhn commented:
I fully agree that if we add type hints that we should do so consistently. `evm/vm/forks/sharding/collations.py` was not supposed to be there. I moved it to `evm/rlp/collations.py` but missed to clean up afterwards. I deleted it now, thanks for spotting this!
